### PR TITLE
Third fase algorithm and interface

### PR DESCRIPTION
**Pull request by BrunoRosendo**
There's currently an exceptions being thrown at the end of the program, related with the erase() function (from vector) dereferencing pointers. I suggest we use references instead of pointers, since it serves the same purpose and it's generally safer and easier to use

### CODE REVIEW

**Comment by BrunoRosendo** [comment]
This concludes this PR. I'm going to continue in a different one

File: src/bakery/Bakery.cpp [hunk 1]
```diff
@@ -188,17 +188,143 @@ void Bakery::greedyWithDijkstra(Van& van) {
...
+
+    if (optimize) {
+        /* Since the last used van probably has space left,
+         * it will try to take some clients from other vans
+         */
+        for (int i = vans.size() - 1; i >= 0; --i)
+            if (!vans[i].getClients().empty()) {
+                for (int j = i - 1; j >= 0; --j) {
+                    Client *client = vans[j].removeFarthestClientInRange(vans[i].getReservedBread());
+                    if (client != NULL) vans[i].addClient(client);
```
Comment (monkin77) [inline src/bakery/Bakery.cpp:289]: Maybe we could have a calculateSimulation function that would check whether the change improves the overall time. That way we could add more than 1 client from the vans that are full. Maybe we do this on another pr tho

File: src/bakery/Bakery.cpp [hunk 2]
```diff
@@ -188,17 +188,143 @@ void Bakery::greedyWithDijkstra(Van& van) {
...
+
+    int clientsAllocated = 0;
+
+    if (useKnapsack) {
+        // Value = breadQuantity / 10 + 1 (the client itself has value)
+        vector<int> values;
+        for (Van &v : vans) {
+            if (clientsAllocated == clients.size()) break;
+
+            for (Client *client : clients)
```
Comment (monkin77) [inline src/bakery/Bakery.cpp:267]: Are the clients sorted by position? By latitude or something of the sort. Since that way each van would have the clients closer together

**Review by monkin77 — APPROVED** [review]
Looks great!

File: src/bakery/Bakery.cpp (hunk 2 above)
Comment (BrunoRosendo) [reply src/bakery/Bakery.cpp:267]: That's an interesting idea but it also doesn't seem that easy. Even if clients have the same latitude, they can still be very far away. But I put a TODO note so we explore this later
